src/cryptonote_basic/miner.cpp

```diff
@@ -846,40 +846,89 @@ namespace cryptonote
 
     #elif defined(__linux__)
 
-      // i've only tested on UBUNTU, these paths might be different on other systems
-      // need to figure out a way to make this more flexible
-      std::string power_supply_path = "";
-      const std::string POWER_SUPPLY_STATUS_PATHS[] = 
-      {
-        "/sys/class/power_supply/ACAD/online",
-        "/sys/class/power_supply/AC/online",
-        "/sys/class/power_supply/AC0/online",
-        "/sys/class/power_supply/ADP0/online"
-      };
+      // Use the power_supply class http://lxr.linux.no/#linux+v4.10.1/Documentation/power/power_supply_class.txt
+      std::string power_supply_class_path = "/sys/class/power_supply";
 
-      for(const std::string& path : POWER_SUPPLY_STATUS_PATHS)
+      boost::tribool is_charging = boost::logic::tribool(boost::logic::indeterminate);
+      boost::tribool on_battery = boost::logic::tribool(true);
+      if (boost::filesystem::is_directory(power_supply_class_path))
       {
-        if( epee::file_io_utils::is_file_exist(path) )
+        const boost::filesystem::directory_iterator end_itr;
+        for (boost::filesystem::directory_iterator iter(power_supply_class_path); iter != end_itr; ++iter)
         {
-          power_supply_path = path;
-          break;
+          const boost::filesystem::path & power_supply_path = iter->path();
+          if (boost::filesystem::is_directory(iter->path()))
+          {
+            std::string power_supply_present;
+            epee::file_io_utils::load_file_to_string((power_supply_path / "/present").string(), power_supply_present);
+            if (!boost::starts_with(power_supply_present, "1")) {
+              LOG_PRINT_L4("Power supply not present at " << power_supply_path);
+              continue;
+            }
+
+            boost::filesystem::path power_supply_type_path = power_supply_path / "/type";
+            if (boost::filesystem::is_regular_file(power_supply_type_path))
+            {
+              std::string power_supply_type;
+              epee::file_io_utils::load_file_to_string(power_supply_type_path.string(), power_supply_type);
+
+              // If there is an AC adapter that's present and online we can break early
+              if (boost::starts_with(power_supply_type, "Mains"))
+              {
+                boost::filesystem::path power_supply_online_path = power_supply_path / "/online";
```

Comment:
As you are now using the / operator I think you are supposed to drop the forward slash in "/online". The / operator add's in the correct path separator.